src/cavendish_particle_tracks/_widget.py

```diff
@@ -123,7 +123,22 @@ def __init__(self, napari_viewer: napari.viewer.Viewer):
         self.decay_angles_isopen = False
 
     def hideEvent(self, event):
-        """When the widget is 'closed' (napari just hides it), show the layer buttons again."""
+        """When the widget is 'closed' (napari just hides it), show the layer buttons again.
+        If data has been recorded, prompt the user to save it before closing the widget.
+        """
+        if len(self.data) > 0:
+            message_box = QMessageBox(self)
+            message_box.setIcon(QMessageBox.Warning)
+            message_box.setText(
+                "Closing Cavendish Particle Tracks. Any unsaved data will be lost."
+            )
+            message_box.setInformativeText("Do you want to save your data?")
+            message_box.setStandardButtons(QMessageBox.Yes | QMessageBox.No)
+            message_box.setDefaultButton(QMessageBox.Yes)
+            reply = message_box.exec()
+
+            if reply == QMessageBox.Yes:
+                self._on_click_save()
```

Comment:
Marginal comment, I might have written the logic a bit differently, and put the messagebox code into a reusable function...
```python
        there_is_data = len(self.data) > 0
        if there_is_data:
            _confirm_save_before_closing()

            if reply == QMessageBox.Yes:
                self._on_click_save()
```

And then have a member function:

```python
def _confirm_save_before_closing(self):
    # Create a confirmation message box.
    message_box = QMessageBox(self)
    message_box.setIcon(QMessageBox.Warning)
    message_box.setText(
        "Closing Cavendish Particle Tracks. Any unsaved data will be lost."
    )
    message_box.setInformativeText("Do you want to save your data?")
    message_box.setStandardButtons(QMessageBox.Yes | QMessageBox.No)
    message_box.setDefaultButton(QMessageBox.Yes)

    # Get users' reply
    reply = message_box.exec()

     if reply == QMessageBox.Yes:
         self._on_click_save()
```

I am wondering if we might want to patch into the windows' close somehow?

Reply:
Modified the logic. How to access the "close" event instead of just "hide" I am not sure yet, but I think we want this protection in nevertheless. New enhancement issue:
- #139